network/retrieve/retrieve_test.go

```diff
@@ -0,0 +1,398 @@
+	bucketKeyNetstore  = simulation.BucketKey("netstore")
+
+	hash0 = sha3.Sum256([]byte{0})
+)
+
+func init() {
+	flag.Parse()
+
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(os.Stderr, log.TerminalFormat(false))))
+}
+
+// TestChunkDelivery brings up two nodes, stores a few chunks on the first node, then tries to retrieve them through the second node
+func TestChunkDelivery(t *testing.T) {
+	chunkCount := 10
+	filesize := chunkCount * 4096
+
+	sim := simulation.NewBzzInProc(map[string]simulation.ServiceFunc{
+		"bzz-retrieve": newBzzRetrieveWithLocalstore,
+	})
+	defer sim.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	defer cancel()
+	_, err := sim.AddNode()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		nodeIDs := sim.UpNodeIDs()
+		log.Debug("uploader node", "enode", nodeIDs[0])
+
+		item := sim.MustNodeItem(nodeIDs[0], bucketKeyFileStore)
+
+		//put some data into just the first node
+		data := make([]byte, filesize)
+		if _, err := io.ReadFull(rand.Reader, data); err != nil {
+			t.Fatalf("reading from crypto/rand failed: %v", err.Error())
+		}
+		refs, err := getAllRefs(data)
+		if err != nil {
+			return err
+		}
+		log.Trace("got all refs", "refs", refs)
+		_, wait, err := item.(*storage.FileStore).Store(context.Background(), bytes.NewReader(data), int64(filesize), false)
+		if err != nil {
+			return err
+		}
+		if err := wait(context.Background()); err != nil {
+			return err
+		}
+
+		id, err := sim.AddNode()
+		if err != nil {
+			return err
+		}
+		err = sim.Net.Connect(id, nodeIDs[0])
+		if err != nil {
+			return err
+		}
+		nodeIDs = sim.UpNodeIDs()
+		if len(nodeIDs) != 2 {
+			return fmt.Errorf("wrong number of nodes, expected %d got %d", 2, len(nodeIDs))
+		}
+
+		// allow the two nodes time to set up the protocols otherwise kademlias will be empty when retrieve requests happen
+		time.Sleep(50 * time.Millisecond)
+		log.Debug("fetching through node", "enode", nodeIDs[1])
+		ns := sim.MustNodeItem(nodeIDs[1], bucketKeyNetstore).(*storage.NetStore)
+		ctr := 0
+		for _, ch := range refs {
+			ctr++
+			_, err := ns.Get(context.Background(), chunk.ModeGetRequest, storage.NewRequest(ch))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+}
+
+// TestDeliveryForwarding tests that chunk delivery forwarding requests happen. It creates three nodes (fetching, forwarding and uploading)
+// where po(fetching,forwarding) = 1 and po(forwarding,uploading) = 1, then uploads chunks to the uploading node, afterwards
+// tries to retrieve the relevant chunks (ones with po = 0 to fetching i.e. no bits in common with fetching and with
+// po >= 1 with uploading i.e. with 1 bit or more in common with the uploading)
+func TestDeliveryForwarding(t *testing.T) {
```

Comment:
what is there not to understand? `findPeer` needs a certain topology for the retrieve requests to propagate through the forwarding node. i'm not which other explanations you expect and what is exactly unclear? the comment, the code? please be a bit more specific

Reply:
also, to further clarify, you cannot test retrieval in a snapshot without pull syncing (since due to the logic in `findPeer` you'll never get to the content unless you build the topology in such a way that every request to every chunk will result in it arriving to the storing node).
so, to test if retrieve requests are being forwarded, a basic topology that adheres to certain POs between nodes is needed, with discovery disabled. This test tests exactly this, just retrieve request forwarding without any other moving parts